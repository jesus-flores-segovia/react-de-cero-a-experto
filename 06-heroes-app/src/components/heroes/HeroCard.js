import React from 'react'
import { Link } from 'react-router-dom'

export const HeroCard = ({
    id,
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters
}) => {

    return (
        <div class="col-sm-3 mb-4">
            <div class="card">
                <img src={`./assets/heroes/${id}.jpg`} className="card-img-top" alt={superhero}/>
                <div class="card-body">
                    <h5 className="card-title">{superhero}</h5>
                    <p className="card-text">{alter_ego}</p>

                    {
                        (alter_ego !== characters)
                            && <p className="card-text">{characters}</p>
                    }

                    <p className="card-text">
                        <small className="text-muted">{first_appearance}</small>
                    </p>

                    <Link to={`./hero/${id}`}>
                        See more
                    </Link>
                </div>
            </div>
        </div>
    )
}
