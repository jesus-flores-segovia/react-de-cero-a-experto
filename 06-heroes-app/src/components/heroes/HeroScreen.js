import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router';
import { getHeroById } from '../../selectors/getHeroById';

const heroes = require.context("../../assets/heroes", true);

export const HeroScreen = ({history}) => {

    const {heroId} = useParams();

    const hero = useMemo(() => getHeroById(heroId), [heroId]);

    if(!hero) { return <Redirect to="/"/>};

    const {
        id,
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;

    const handleReturn = () => history.length <= 2 ? history.push("/") : history.goBack();

    return (
        <div className="row">
            <div className="col-4 animate__animated animate__fadeInLeft">
                <img src={heroes(`./${id}.jpg`).default} className="img-thumbnail" alt={superhero} />
            </div>
            <div className="col-8">
                <h3>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"><b>Alter ego: </b>{alter_ego}</li>
                    <li className="list-group-item"><b>Publisher: </b>{publisher}</li>
                    <li className="list-group-item"><b>First Appearance: </b>{first_appearance}</li>
                </ul>

                <h5 className="mt-3">Characters</h5>
                <p>{characters}</p>

                <button 
                    className="btn btn-primary"
                    onClick={handleReturn}
                >
                Go back
                </button>
            </div>
        </div>
    )
}
