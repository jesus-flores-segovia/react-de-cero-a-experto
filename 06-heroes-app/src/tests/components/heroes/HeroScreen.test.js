import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { HeroScreen } from "../../../components/heroes/HeroScreen"

describe("Tests inside file 'HeroScreen.test.js'", () => {

    const history = {
        length: 0,
        push: jest.fn(),
        goBack: jest.fn(),
    };

    test("HeroScreen component must be shown as empty because it have not a valid hero url and redirects to the login screen", () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={["/hero"]}>
                <HeroScreen history={history}/>
            </MemoryRouter>
        );

        expect(wrapper.find("Redirect").exists()).toBeTruthy();
    })

    test("HeroScreen component must be shown a hero if the url parameter exists and it's found", () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
                <Route path="/hero/:heroId" component={HeroScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".row").exists()).toBeTruthy();

    })

    test("HeroScreen component must return to the login page when press the 'Go back' button if the user access to the hero screen between the directly url", () => {
        
        const history = {
            length: 2,
            push: jest.fn(),
            goBack: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={history}/>} />
            </MemoryRouter>
        );

        wrapper.find("button").prop("onClick")();
        expect(history.push).toBeCalledTimes(1);
        expect(history.push).toBeCalledWith("/");
        expect(history.goBack).not.toBeCalled();

    })

    test("HeroScreen component must return to the hero card list screen when press the 'Go back' button if the user complete the full hero journey", () => {
        
        const history = {
            length: 3,
            push: jest.fn(),
            goBack: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={["/hero/marvel-spider"]}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={history}/>} />
            </MemoryRouter>
        );

        wrapper.find("button").prop("onClick")();
        expect(history.goBack).toBeCalledTimes(1);
        expect(history.push).not.toBeCalled();

    })

    test("HeroScreen component must be shown as empty because it have not a valid hero id url and redirects to the login screen", () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={["/hero/marvel-spider-not-exists"]}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={history}/>} />
            </MemoryRouter>
        );

        expect(wrapper.text()).toBe("");
    })
})