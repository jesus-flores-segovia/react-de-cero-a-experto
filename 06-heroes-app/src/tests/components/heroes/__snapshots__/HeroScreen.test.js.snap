// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'HeroScreen.test.js' HeroScreen component must be shown a hero if the url parameter exists and it's found 1`] = `
<div
  className="row"
>
  <div
    className="col-4 animate__animated animate__fadeInLeft"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail"
      src="/assets/heroes/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <b>
          Alter ego: 
        </b>
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <b>
          Publisher: 
        </b>
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <b>
          First Appearance: 
        </b>
        Amazing Fantasy #15
      </li>
    </ul>
    <h5
      className="mt-3"
    >
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Go back
    </button>
  </div>
</div>
`;
